### ISSUE TITLE
GNNExplainer

### ISSUE THREAD

**Issue by paulamartingonzalez** [dmlc/dgl#2787]
## 🚀 Feature
I have just seen that @zhjwy9343 @KounianhuaDu have uploaded some preliminary work for GNNexplainer implementation in DGL. Thanks for such amazing work!

I have seen that the implementation is so far only for node predictions and I was wondering if you are planning to implement the same for graph-level predictions as it would be super helpful. Also, I was wondering if you are still polishing it or if the code is ready for usage!

I am happy to help in any way for this! Many thanks!

**Comment by zhjwy9343** [comment]
@paulamartingonzalez glad you notice this example. 

We don't have plan to implement graph-level prediction in recent time. The example itself is ready for use, and you can refer to its implementation for node classification after modify the to-be-explained model a little bit as I did.

Also, if you like to contribute to this example or even a new example for link- and graph-level explainer, you are well welcomed. If you plan so, I am willing to help on some details.

Thanks again.

**Comment by jermainewang** [comment]
Close for now. @paulamartingonzalez if you're willing to help, feel free to re-open it again.

**Comment by paulamartingonzalez** [comment]
I would be happy to work on the GNN implementation for the graph level features! @zhjwy9343 would you mind pinpointing me to the key details to change? I am still studying the original paper and code. Thanks! 😄

**Comment by zhjwy9343** [comment]
@paulamartingonzalez I just reopened this issue. But, I would like to suggest we use Slack for better communication. There is a Slack channel, named "DGL". 

In terms of Graph-level explanation, the author Rex Ying has described the method as "When classifying a graph, the adjacency matrix in Eq. (5) is the union of adjacency matrices for all nodes in the graph whose label we want to explain. However, note that in graph classification, unlike node classification, due to the aggregation of node embed- dings, it is no longer true that the explanation GS is necessarily a connected subgraph. Depending on application, in some scenarios such as chemistry where explanation is a functional group and should be connected, one can extract the largest connected component as the explanation." 

In the author's code, I can find a Graph level explainer at https://github.com/RexYing/gnn-model-explainer/blob/master/explainer/explain.py#L355. So you might want to start from there and ping me on the Slack channel for further discussion.

**Comment by fmellomascarenhas** [comment]
Hi all, I was checking the code and I couldn't figure out the intuition behind the first term of the loss function

```
#In the paper: 1. The prediction loss between logits before and after node and edge masking;
#In the PyG implementation: 1. The prediction loss between predict logits before and after node and edge masking;
#pred_logits was generated by the explainer.
#label_logits was generated by the original model

log_logit = - F.log_softmax(pred_logits, dim=-1)
pred_loss = th.sum(log_logit * pred_label)
```

Since we are trying to minimize this loss, the model would just always try to find pred_logits = 1, so log_logit would be = -1, and pred_loss would be the minimum possible value.

What is the intuition behind that? I would imagine that a better approach would be to minimize the difference between pred_label and pred_logits, so that we can achieve the same result while pruning the maximum amount of edges and features.

**Comment by zhjwy9343** [comment]
@fmellomascarenhas your guess is right. The pred_loss is computing the cross entropy. 

Here the logits is the outputs of a to-be-explained model. It is normally a N*D tensor, where D is the number of classes. And the pred_label is one-hot encoded before calling the explainer, which is N*D tensor too.

The F.log_softmax will transfer the logits from N*D tensor with real values to softmax probability and then to log values. After log_softmax plus negative, multiple the two N*D tensors to get the cross_entropy.
